src/common/multi_file_reader.cpp

```diff
@@ -180,7 +214,18 @@ void MultiFileReader::FinalizeBind(const MultiFileReaderOptions &file_options, c
 			bool found_partition = false;
 			for (auto &entry : options.hive_partitioning_indexes) {
 				if (column_id == entry.index) {
-					reader_data.constant_map.emplace_back(i, Value(partitions[entry.value]));
+					Value value(partitions[entry.value]);
+					if (!file_options.hive_types_schema.empty()) {
+						auto it = file_options.hive_types_schema.find(entry.value);
+						if (it != file_options.hive_types_schema.end()) {
+							if (!value.TryCastAs(context, it->second)) {
+								throw InvalidInputException(
+								    "Unable to cast '%s' (from hive partition column '%s') to: '%s'", value.ToString(),
+								    StringUtil::Upper(it->first), it->second.ToString());
```

Comment:
I think this block of code is nested a bit too deeply. Could you refactor it to clean it up a bit? Maybe check out the [return early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8) which can help avoid these code structures